Day 22: “Semantics beyond the tag name” / Nathan Knowler (#213)
matuzo/HTMHell: hell/adventcalendar/2025/22/index.md

---
title: "Semantics beyond the tag name"
author: "Nathan Knowler"
author_bio: "HTML and CSS writer living in [Winnipeg](https://www.youtube.com/watch?v=AjqgTtl3YBQ). Passionate about accessiblity, progressive enhancement, and web components. Probably can be found in a mosh pit."
date: 2025-12-22
author_links:
  - label: "Site"
    url: "https://knowler.dev"
    link_label: "knowler.dev"
  - label: "Mastodon"
    url: "https://sunny.garden/@knowler"
    link_label: "@knowler@sunny.garden"
  - label: "Bluesky"
    url: "https://bsky.app/profile/knowler.dev"
    link_label: "@knowler.dev"
intro: "<p>Semantic HTML is the cure to divitis, but we need to go beyond tag names if we want to truly understand how to write it.</p>"
image: "advent25_22"
---

There is a terrible epidemic that plagues the web: <i>divitis</i>.

> If you’re unfamiliar, <dfn><i>divitis</i></dfn> is a condition where a website uses an unholy amount of `<div>` elements. Some of the worst cases even include `<div>`s used as buttons—_blasphemy!_

Reviewer: > `<div>`s used as buttons
😲

Reviewer 2: 🥴

```html
<div class="site-header">
  <div class="site-title">divs r us</div>
  <div class="navigation">
    <div class="link">my favourite divs</div>
    <div class="link">moor divs</div>
  </div>
</div>
<div class="main">
  <div class="content">
    <div class="container">
      <div class="container-inner">
        <div class="card">
          <div class="card-inner">
            <div class="card-header">
              <div>i love div</div>
            </div>
            <div>all my friends are divs</div>
            <div class="card-footer">
              <div class="button">accept div as your lord and saviour</div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
<div class="site-footer">
  <div class="div-webring">
    <div class="button">join the divitis webring</div>
    <div class="button next">next</div>
    <div class="button prev">previous</div>
  </div>
</div>

> ```

Reviewer: It's fine for me.

> ```

Reviewer 2: Agree it's fine, making it a bit OTT is helpful to the point here I think?

> To avoid <i>divitis</i>, authors are often encouraged to use <i>semantic HTML</i>. As it turns out, HTML has over a hundred elements that are not named `div` or `span`. Semantic HTML elements describe what their content is.

Reviewer 2: Take this as very much of a side note, but in my experience an equally (and perhaps even bigger) cause of divitis is bloat and unnecessary wrappers (often due to insufficient CSS knowledge). Even if authors suddenly used all the right semantic elements, if you squint, the output would still be mostly divs. 

Up to you if worth mentioning (even just replacing with "_One of the ways_ of avoiding..."), just thought I'd note!

Author: I didn’t end up getting to this and probably won’t given the time, but I do agree!

> Inspired by this newfound responsibility, authors open the lexicon of HTML elements and begin to compose… _absolute gibberish._

Reviewer 2: It's fairly obvious to a certain audience, but since you don't immediately go on to explain why the given example is gibberish, it may leave less knowledgeable readers feeling lost 🥹 Maybe just as a kindness (and also to improve signposting), something along the lines of "(we’ll unpack why this structure doesn’t mean what the author thinks it means in a moment)"...  just a small cue so less experienced readers don’t feel talked down to.

Or, after the example:

> While it can look deceptively semantic at first (no divs!), in this case it doesn't mean what the author thinks it means.

Author: Thinking out loud (and leaving this as a note for myself mostly), I should make better use of this example throughout the post like you suggest elsewhere. I was a bit lazy with including this example in the first place (i.e. I couldn’t think of anything on my own, so I went through the site to find one).

Author: Should be resolved in [ffaa2dd](https://github.com/matuzo/HTMHell/pull/213/commits/ffaa2dd8aebcad201afa5fc38b9a49e92a8e66ee)

Reviewer 2: That, and the new example mirroring the divitis one with the internal monologue is a really clever way to make it obvious 😆 love it

<!-- Copied from Hell #29 -->

```html
<section>
  <aside>
    <div>
      <section>
        <header>
          <a href="/">
            <img src="logo.svg" alt="Logo">
          </a>
        </header>
        <main>
          <a href="/services">Services</a>
          <a href="/products">Products</a>
          <a href="/aboutus">Aboutus</a>
        </main>
        <footer></footer>
      </section>
    </div>
  </aside>
  <section>
    <footer></footer>
    <main>
      <h1>Welcome to Hell</h1>
    </main>
    <footer></footer>
  </section>
</section>
```

> To be fair, this is to be expected of anyone learning a new language, however, a problem on the web is that authors delude themselves into thinking that they are “writing semantic HTML,” therefore gaining all of the benefits that are understood to come with that like accessibility, and they never advance past that stage.

Reviewer 2: Just a pause to shorten the sentences:

```suggestion
To be fair, this is to be expected of anyone learning a new language. However, a problem on the web is that authors delude themselves into thinking that they are “writing semantic HTML,” therefore gaining all of the benefits that are understood to come with that like accessibility, and they never advance past that stage.
```

Author: Resolved in [365d460](https://github.com/matuzo/HTMHell/pull/213/commits/365d4600d52b678bb8e4a099bd20b132142648f7)

It’s like when my own child confidently declares, after learning a handful of words in Japanese, “I know Japanese.” It’s cute and I’m not going to argue with it, because they’re a child.

You’re likely not a child though, so I will tell it to you straight: _this is not cute._

> Once you’ve recovered from that devestating realization, we can proceed.

Reviewer: A little bit maybe, but imo it's not too much. We'll see what Elise has to say. :)

> Once you’ve recovered from that devestating realization, we can proceed.

Reviewer 2: I didn't feel that way when first reading through it, but if you're worried, maybe my comment above about explaining why it's gibberish will make this land softer.

To actually write semantic HTML, we need to know what elements mean beyond just what we infer from their tag names and how to use them. So, how do we find that out?

## Where HTML semantics come from

[The HTML standard][html] is what defines HTML elements. For example, we can find the definition of the `<main>` element in [section 4.4.14](https://html.spec.whatwg.org/multipage/grouping-content.html#the-main-element):

> The `main` element represents the dominant contents of the document.

Beyond the definition, the standard also includes rules and guidance for usage:

> A document must not have more than one `main` element that does not have the `hidden` attribute specified.

> A <dfn>hierarchically correct `main` element</dfn> is one whose ancestor elements are limited to `html`, `body`, `div`, `form` without an accessible name, and autonomous custom elements. Each `main` element must be a hierarchically correct `main` element.

> From these descriptions, we can gather that using the `<main>` element for the main content of a card component that’s meant to be used many times throughout a document would be incorrect:

Reviewer 2: Maybe it'd feel more coherent if instead of giving a new example, you could use this as a chance to reference back to the first gibberish example and explain why that doesn't work?

Author: In [6389673](https://github.com/matuzo/HTMHell/pull/213/commits/6389673db79283331e16516f0500880bb4775615), I changed the original gibberish example, then in [91dde92](https://github.com/matuzo/HTMHell/pull/213/commits/91dde925b63119d889fd9d7873832c887efc8f45) I use a truncated/focused version of that to point out the issue.

```html
<div class=card>
  <main>
    First card
  </main>
</div>
<div class=card>
  <main>
    Second card
  </main>
</div>
```

Tag names are not enough for us to understand what an element is for. We need to adopt HTML’s own definitions for semantic elements.

### Why follow the HTML Standard?

> Now, I’m not your god and neither is the HTML Standard. You don’t have to follow it if you don’t want to. If you choose to use HTML elements for how you perceive them, there will more than likely be undesirable consequences that you cannot control. Consequences like assistive technologies not working well with your documents which can also lead to legal consequences if there are accessibility laws where you live. Who knows? If you don’t follow the HTML Standard, maybe your documents might just end up [in hell](https://www.htmhell.dev/).

Author: Similar to my comment on lines 95-101, this seems like it also could be patronizing. I am trying to play off the theme of the site (i.e. “hell”).

Reviewer: Fine for me.

Reviewer 2: Felt fine to me on first read-through as well, but the thought is perhaps a bit long-winded, and that can also make it feel more preachy than it needs to be?

Suggest to keep the humour but just tighten it a bit:

```suggestion
I’m not your god, and neither is the HTML Standard – but if you choose to use HTML elements for how you perceive them, things break in ways you can’t control. Assistive tech, legal requirements, user expectations... it all hinges on the platform’s actual definitions. Ignore them and, well, your documents might just end up [in hell](https://www.htmhell.dev/).
```

Author: @elisehein That reads so much better. Thanks! Resolved in [365d460](https://github.com/matuzo/HTMHell/pull/213/commits/365d4600d52b678bb8e4a099bd20b132142648f7)

### Some elements are “more semantic” than others

Something that’s important to understand about semantic HTML is that, while HTML may define all of its elements, not all elements are meaningful for assistive technologies. With that said, you can spend a lot of time writing valid, semantically correct HTML and accomplish absolutely nothing for what makes semantic HTML the most valuable: [making the web for everyone](https://www.w3.org/TR/w3c-vision/#vision-web).

> Beyond the tag name and even HTML semantics, there’s another invisible layer of semantics that we need to understand to write semantic HTML well: implicit ARIA semantics.

Reviewer 2: Unsure if this is the best place for it, but for readers less familiar with the topic, I think calling out that there are also _explicit_ semantics (eg `role="navigation"`), and that they always override the implicit role, would help complete the picture.

Author: I think the changes I made in [1e45ef1](https://github.com/matuzo/HTMHell/pull/213/commits/1e45ef1ee142ed4947f10263b8307efcd063c557) mostly resolve your concern here since it front-loads introducing ARIA as a concept through the use of explicit attributes. It doesn’t mention anything about explicit ARIA overrides implicit ARIA. If time permits, I might try to work that in somewhere, it could even be a highlighted note.

> 

Reviewer: This is confusing. For most people, writing semantic HTML means writing HTML that works well for AT. If there are two types of semantics, why would I need the one that doesn't provide anything for AT? Why does it even exist? What are the other benefits of using semantic elements besides implicit ARIA roles?
You're explaining ARIA semantics in detail, that's fine, but it's not clear how you differentiate between HTML semantics and ARIA semantics, and what exactly HTML semantics do and what we need them for.

Reviewer 2: I think maybe this is just a question of wording/positioning? The way "implicit ARIA semantics" it’s introduced here can read as if HTML semantics and ARIA semantics are separate systems, or two kinds of semantics authors need to choose between.

Reviewer 2: Maybe literally just expanding on the definition of "ARIA semantics" would help – the role, properties/states, accessibility mappings etc _derived from HTML semantics_ (rather than an unrelated other system)

Author: Thinking out loud (again), I think what I’m trying to accomplish here and with this post is to steer readers not only towards using semantic HTML elements, but semantic HTML elements that are impactful for good accessibility. Perhaps a better problem to point out here is: there are a lot of semantic HTML elements — how do we know which ones to prioritize for good accessibility? Then I can answer with introducing ARIA semantics, focusing specifically on the implicit ones (but call out explicit ones since the reader has a better chance at being familiar with them).

Author: In https://github.com/matuzo/HTMHell/pull/213/commits/1e45ef1ee142ed4947f10263b8307efcd063c557, I decided to introduce ARIA first, rebutting the idea that one might use ARIA instead of semantic HTML and then using this as an opportunity to introduce semantic HTML’s implicit ARIA which I hope avoids setting them up as competing semantics. Since I’ve done this, I slightly adjusted the title of the next section to reflect the change of where the ARIA definition is, since that section now only focuses on how implicit ARIA works in HTML.

Reviewer 2: Love this direction!

## Implicit ARIA semantics

You may be familiar with [ARIA (Accessible Rich Internet Applications)][aria]. ARIA is a set of roles and attributes that can be used in HTML to add accessibility related semantics to web content.

> [The first rule of ARIA usage](https://www.w3.org/TR/using-aria/#firstrule) is: don’t use ARIA if you don’t need to. Why? Because many semantic HTML elements and attributes have what’s called implicit ARIA semantics—as well as the built-in functionality necessary for those semantics.

Reviewer 2: I get the point here, but the phrasing threw me for a moment because just two paragraphs ago you were saying it's NOT enough to just use the right tag names.

Author: I believe my reworking of the prior section resolves this concern: [1e45ef1](https://github.com/matuzo/HTMHell/pull/213/commits/1e45ef1ee142ed4947f10263b8307efcd063c557)

### Attributes affect implicit semantics

Consider the following elements:

```html
<a>Anchor</a>
<a href=https://example.com>Anchor with an <code>href</code> attribute</a>
```

Are these both links?

No, they aren’t. The first is just an anchor element. It’s implicit ARIA role is `generic` which is not interesting for accessibility APIs. CSS also doesn’t even consider it a link as it will not match the `:any-link` or `:link` pseudo-class selectors.

Since the latter anchor element has an `href` attribute, it is considered a link, and therefore has the implicit ARIA role of `link` and can be matched by the above mentioned semantic pseudo-class selectors with CSS.

### Naming affects implicit semantics

A popular favourite among those first starting out with semantic HTML is the `<section>` element. While [the abuses of this element can be extensive](https://www.htmhell.dev/10-section-is-no-replacement-for-div/), what is generally never realized is that it is semantically insignicant if it is not named.

```html
<section>
  <h2>Insignificant section</h2>
</section>
<section aria-labelledby=section-label>
  <h2 id=section-label>Significant document region</h2>
</section>
```

> With that said, please do not go name _all_ of your `<section>` elements! If they were meant to be named, then the browser would do this for you. Often adding more landmarks makes it more difficult to identify what’s important on a page.

Reviewer: > If they were meant to be named, then the browser would do this for you.

I don't understand this sentence. The nav or dialog elements are supposed to be named, but the browser still doesn't do it for me.

Reviewer: > Often adding more landmarks

You are referring to landmarks, but you're using the term _landmark_ here for the first time.

Author: I can see how this was confusing. I expanded my reasoning here and formally introduced the term landmark for the reader in https://github.com/matuzo/HTMHell/pull/213/commits/2bd554d4e2c5aca089118a07911d5fba76b6ccde. I might need to do some more tweaking as I think it’s kind of odd that I call out that `<section>` elements have no ARIA semantics when not named, but then immediately advise against adding names for most cases.

Author: In [eb16470](https://github.com/matuzo/HTMHell/pull/213/commits/eb164708316b72aefe8192fd41de55e6d1af5553), I made some final adjustments to this section. I now focus on the fact that most `<section>` usage is no different than using a `<div>` and their implicit roles are the same. I think that and explaining the `region` role and landmarks makes it more clear why folks shouldn’t name all of their sections.

> With that said, please do not go name _all_ of your `<section>` elements! If they were meant to be named, then the browser would do this for you. Often adding more landmarks makes it more difficult to identify what’s important on a page.

Reviewer 2: I think this lacks some explanation – it first says they're semantically insignificant without a name, but then says not to name them.

Author: As noted in https://github.com/matuzo/HTMHell/pull/213#discussion_r2532933128, I’ve begun trying to improve this section with an expanded explanation. I will try to further address this concern though.

```html
<body>
  <header>Header content</header>
  <main>Main content of the document</main>
  <footer>Footer content</footer>
</body>
```

In this scenario, the `<header>` element’s implicit ARIA role is `banner` and the `<footer>` element’s implicit ARIA role is `contentinfo`. Documents should only have one of each of these landmarks.

That does not mean that you cannot use both of these elements in other places within your document, however, if they contain content that should be described by either of those roles, then you likely need to restructure your markup.

As of 2025, when used inside of the `<main>` element or [sectioning content](https://html.spec.whatwg.org/multipage/dom.html#sectioning-content), `<header>` and `<footer>` elements have been mapped to new ARIA roles: [`sectionheader`](https://w3c.github.io/aria/#sectionheader) and [`sectionfooter`](https://w3c.github.io/aria/#sectionfooter).

```html
<body>
  <header>banner landmark role</header>
  <main>
    <article>
      <header>sectionheader role</header>
      <p>Article content</p>
      <footer>sectionfooter role</footer>
    </article>
  </main>
  <footer>contentinfo landmark role</footer>
</body>
```

> As of writing, this change has only been implemented in Chromium and WebKit-based browsers. [Gecko has yet to implement them](https://bugzilla.mozilla.org/show_bug.cgi?id=1893684). As these are new roles, the support from assistive technologies will likely be non-existent or quite limited.

Reviewer: It's worth noting that, before this addition, the implicit role of the header and footer outside the body context was/is _generic_.

Author: [db2fce2](https://github.com/matuzo/HTMHell/pull/213/commits/db2fce204e7bb9ab1b38fc1071f5392854601e6b) Ok, so I think I improved this section a fair bit. Now, I do explicit mention that they were `generic` prior.

> As of writing, this change has only been implemented in Chromium and WebKit-based browsers. [Gecko has yet to implement them](https://bugzilla.mozilla.org/show_bug.cgi?id=1893684). As these are new roles, the support from assistive technologies will likely be non-existent or quite limited.

Reviewer: I'm not sure. As far as I know, they're still debating about whether and how to implement these roles.

### Custom elements can have implicit semantics too

[Custom elements][html-custom-elements] are an HTML feature that allow authors to define their own HTML elements. A tag name can be defined as a custom element using the `window.customElements.define()` method in JavaScript. Like built-in HTML elements, custom elements can have their own implicit ARIA semantics. This is set using the [`ElementsInternals`](https://developer.mozilla.org/en-US/docs/Web/API/ElementInternals) API.

Here is an example of defining a custom element and setting an implicit ARIA role for it:

```html
<script>
  class HellButtonElement extends HTMLElement {
    constructor() {
      super();

      const internals = this.attachInternals();

      internals.role = "button";

      // Continue implementing button functionality 🙈
    }
  }

  customElements.define("not-a-div", HellButtonElement);
</script>

<not-a-div>A div is not a button</not-a-div>
```

When you create custom elements, you become responsible for communicating their purpose to other authors. This should also include any implicit semantics you’ve set, as well as, what you consider good authoring practice to be when using them.

> I highly recommend using HTML itself as a model for how you build custom elements. In many ways, that might mean thinking less like an author and more like an implementor or a specification writer. “How do I create elements for others to use?” Implicit semantics are an important part of that because they allow you to [put other authors first](https://www.w3.org/TR/design-principles/#priority-of-constituencies).

Reviewer 2: Love this part

## Test it in a browser

In the same way that you can spend a whole lot of time writing valid, semantic HTML that has no implicit ARIA semantics, you can also spend a whole lot of time on writing HTML according to the specifications that actually isn’t implemented in any browser or that is implemented incorrectly in browsers. This is why testing is vital.

Testing might look like checking what ARIA roles and properties are computed for elements in a browser’s developer tools. It also might look like actually using assistive technologies such as screen readers or voice control. It is very helpful to understand how assistive technology users experience various elements.

> Testing helps us understand what is valuable and what we should focus on. I can spend a whole lot of time trying to perfectly mark up all of my words using [text-level semantics][html-text-level-semantics], but then realize through testing that I’ve delivered little to no semantic value to assistive technology users.

Reviewer 2: 100% agree, but in my experience, testing with real AT is considered a huge barrier to most devs. They're unfamiliar territory if you don't use AT yourself, and often feel like they're gonna be really tricky to master. The reality is that very few devs ever try it. 

Not sure where I'm going with this tbh, but it's just a thought I found myself having as I was reading this bit – kinda like, huh, well _that's_ wishful thinking...

Maybe a little note of encouragement or links to resources (eg Sara Soueidan's Practical Accessibility course?) would make this feel more approachable?

Author: This was a good callout. I added some links to how to find the accessibility info in developer tools and added a couple of course recommendations in [08191b0](https://github.com/matuzo/HTMHell/pull/213/commits/08191b0683115716b5e33e8fb06937f92ff6ac84).

## Go forth and actually write semantic HTML

We’ve learned that:

1. HTML tag names are not a reliable way of understanding what an element is for.
2. The HTML standard defines what semantic HTML elements are for.
3. Some semantic HTML elements also have implicit ARIA semantics, and in some cases that depends on what attributes are set, if the element has an accessible name, the context the element is used in, and even a combination of these things.
4. Custom elements can have implicit ARIA semantics too.
5. Don’t take anyone’s word for it: test it yourself.

> 

Reviewer 2: Nice 🥇

  <div class="div-webring">
    <div class="button">join the divitis webring</div>
    <div class="button next">next</div>
    <div class="button prev">previous</div>
  </div>
</div>
```

To avoid <i>divitis</i>, authors are often encouraged to use <i>semantic HTML</i>. As it turns out, HTML has over a hundred elements that are not named `div` or `span`. Semantic HTML elements describe what their content is.

Inspired by this newfound responsibility, authors open the lexicon of HTML elements and begin to compose… _absolute gibberish._

```html
<my-card>
  <header>
    <section>
      <h1>divs suck</h1>
      <h2>and so do classes</h2>
    </section>
  </header>
  <main>
    <section>who needs classes when you have semantic HTML. combine elements however you’d like to explore new possibilities.</section>
  </main>
  <footer>
    <section>
      <nav>
        <a href="/hell">
          <button>
            <label>I have no class</label>
          </button>
        </a>
      </nav>
    </section>
  </footer>
</my-card>

<my-card>
  <header>
    <section>
      <h1>replace all your divs with sections</h1>
      <h2>they’re so semantic</h2>
    </section>
  </header>
  <main>

>     <section>whenever I don’t know what element to choose, section is there for me. section is my best friend forever.</section>

Reviewer 2: 🤣

```html
<div role=button>accept div as your lord and saviour</div>
```

Many semantic HTML elements have what’s called <i>implicit ARIA semantics</i>. These are ARIA roles and properties that the browser has mapped for us. Instead of needing to explicitly write out the ARIA semantics for an element each time you use it, we can use semantic HTML elements as a shorthand for these.

```html
<button>This machine kills fascists</button>
```

Beyond simply being a shorthand for a set of ARIA semantics, built-in HTML elements also include the necessary functionality and presentation that would be expected of an element bearing such semantics. In many cases, this means that you don’t need to write any JavaScript for your HTML to work as expected. Using JavaScript in lieu of built-in HTML functionality is wasteful and error prone, degrading the usability of your website.

> Now that we understand the folly of ignoring the first rule of ARIA usage, let’s turn our attention to <i>implicit ARIA semantics</i> and how they work in semantic HTML.

Reviewer 2: Nice 👌

```html
<section>
  <h2>Insignificant section</h2>
</section>
<section aria-labelledby=section-label>
  <h2 id=section-label>Significant document region</h2>
</section>
```

With that said, please do not go name _all_ of your `<section>` elements!

```html
<section aria-labelledby=section-1-label>
  <h1 id=section-1-label>Top one hundred reason I love sections</h1>
  <p>…</p>
</section>
<section aria-labelledby=section-2-label>
  <h2 id=section-2-label>Section is whatever I want it to be</h2>
  <p>…</p>
</section>
<section aria-labelledby=section-3-label>
  <h2 id=section-3-label>Section is my friend</h2>
  <p>…</p>
</section>
```

A labelled `<section>` element will have [the `region` ARIA role][aria-region]. This is what’s called a <dfn><i>landmark</i></dfn>. [Landmarks](https://tetralogical.com/blog/2022/03/18/landmarks/) are important sections of a document that a user might benefit from easy access to.

The problem here is that everything is a landmark, then nothing is a landmark.

> It can be easy to think that labelling your `<section>` elements filling a requirement for using them, but that is not the case here. Both the meaning of the tag name and the HTML definition of a `<section>` element is way more generic than the `region` ARIA role. Before you label your `<section>`  elements ensure they make sense as regions, otherwise, you’ll likely be making it more difficult to find what’s important on a page. If you’re already using [a good heading structure](https://tetralogical.com/blog/2022/02/28/headings/) throughout your document, that is good enough for what you’re likely trying to do.

Reviewer 2: I think a word has gone missing here

```suggestion
It can be easy to think that labelling your `<section>` elements satisfies the requirement for using them, but that is not the case here. Both the meaning of the tag name and the HTML definition of a `<section>` element is way more generic than the `region` ARIA role. Before you label your `<section>`  elements ensure they make sense as regions, otherwise, you’ll likely be making it more difficult to find what’s important on a page. If you’re already using [a good heading structure](https://tetralogical.com/blog/2022/02/28/headings/) throughout your document, that is good enough for what you’re likely trying to do.
```

Reviewer 2: This whole section is much more accessible now!

Also, as a complete aside, irrelevant to this post, but genuinely interested in what you think – what are your thoughts on the DX vs UX aspects of using HTML tagnames like `<section>` that don't make a difference re literal assistive tech usability, but just make it ✨ nicer ✨ to the devs having to iterate on the code?

As a developer working in a team, I often use HTML elements with an implicit `generic` role literally just to improve readability of code (come to think, I'd prob do this when working on a solo project too). In a real project, the HTML for a single view might be scattered across very many files, so any hints at the structure and intended nesting are useful.

I wonder if there can be any risks to this? I guess it comes down to being mindful of intent.

Author: @elisehein I had this question in the back of mind while writing this and it’s something I’ve changed my mind about in the last year or so. Now, I tend to agree.

When I wrote [the precursor to this article](https://knowler.dev/blog/context-matters-for-semantic-html) a couple of years ago, with regard to `<header>` and `<footer>` elements, I thought that using them outside of contexts where they had (ARIA) semantic value was confusing and it unnecessarily complicated selecting them with CSS (i.e. if you only have one of each of these, then you don’t need to invent a class name or use a selector with more specificity).

At the same time, I had already been using elements with valid custom element tag names instead (e.g. rather than `<article><header>`, I’d use `<article><article-header>`), but that creates a styling burden in another way since the initial `display` value is `inline`—something that is particularly noticeable on [CSS Naked Day](https://css-naked-day.org/)—so I largely stopped doing that for my personal projects, but I’m not opposed to it (this approach is safer for replacing `span` elements).

Now, I think that I’d probably just use elements according to their HTML definition even if they have no ARIA semantic value in certain contexts, unless there was a reason that it was problematic. I definitely use `<section>` a lot more in cases that it doesn’t have any implicit ARIA semantic value—I can only think of two or three cases where I’ve ever associated a label to get `role=region` out of it.

> I wonder if there can be any risks to this? I guess it comes down to being mindful of intent.

I think the whole `sectionheader`/`sectionfooter` change for `<header>`/`<footer>` demonstrates what could happen: if someone had good authoring practices, then they’ll benefit from the semantic upgrade, but if they weren’t used correctly, there could be some unexpected consequence. I presume those making the decisions to add the roles noticed that authoring practices were generally good (the elements are named pretty well).

Maybe another risk: someone who is inexperienced in accessibility comes along and starts to make changes that activate the implicit roles for these elements (or even adds them explicitly). It’s easy to accidentally read [MDN’s “Implicit ARIA role” for `<section>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/section#technical_summary) and conclude “we’re not using `<section>` correctly unless we add an accessible name.”

I do think the risks are generally worth it. Ideally, HTML/HTML-AAM shouldn’t be making breaking changes, so using HTML according to how it defines itself should be safe.

Reviewer 2: Thanks for taking the time to share your thoughts, interesting stuff!

Testing might look like checking what ARIA roles and properties are computed for elements in a browser’s developer tools ([use the accessibility tree in Chrome/Edge and Firefox](https://devtoolstips.org/tips/en/see-accessibility-tree/) or [Accessibility section of the Node Panel in Safari’s Web Inspector](https://webkit.org/web-inspector/elements-tab/#node-panel)). It also might look like actually using assistive technologies such as screen readers or voice control. It is very helpful to understand how assistive technology users experience various elements and so seeking real feedback is invaluable to the testing process. Just make sure to show anyone you do involve that you value their time and labour.

Testing helps us understand what is valuable and what we should focus on. I can spend a whole lot of time trying to perfectly mark up all of my words using [text-level semantics][html-text-level-semantics], but then realize through testing that I’ve delivered little to no semantic value to assistive technology users.

> Courses such as [Sara Soueidan’s Practical Accessibility](https://practical-accessibility.today) or [Marcy Sutton’s Testing Accessibility](https://testingaccessibility.com) are great resources for those wanting to get started with testing various assistive technologies.

Reviewer 2: ❤️
